Lib/test/test_builtin.py

```diff
@@ -1511,6 +1513,126 @@ def test_construct_singletons(self):
             self.assertRaises(TypeError, tp, 1, 2)
             self.assertRaises(TypeError, tp, a=1, b=2)
 
+
+class TestBreakpoint(unittest.TestCase):
+    def test_breakpoint(self):
+        with patch('pdb.set_trace') as mock:
+            breakpoint()
+        mock.assert_called_once()
+
+    def test_breakpoint_with_breakpointhook_set(self):
+        call_status = 'Not called'
+        def my_breakpointhook():
+            nonlocal call_status
+            call_status = 'Called'
+        try:
+            sys.breakpointhook = my_breakpointhook
+            breakpoint()
+        finally:
+            sys.breakpointhook = sys.__breakpointhook__
+        self.assertEqual(call_status, 'Called')
+
+    def test_breakpoint_with_breakpointhook_reset(self):
+        call_status = 'Not called'
+        def my_breakpointhook():
+            nonlocal call_status
+            call_status = 'Called'
+        try:
+            sys.breakpointhook = my_breakpointhook
+            breakpoint()
+        finally:
+            sys.breakpointhook = sys.__breakpointhook__
+        self.assertEqual(call_status, 'Called')
+        with patch('pdb.set_trace') as mock:
+            breakpoint()
+        mock.assert_called_once()
+
+    def test_breakpoint_with_args_and_keywords(self):
+        call_args = None
+        def my_breakpointhook(*args, **kws):
+            nonlocal call_args
+            call_args = args, kws
+        try:
+            sys.breakpointhook = my_breakpointhook
+            breakpoint(1, 2, 3, four=4, five=5)
+        finally:
+            sys.breakpointhook = sys.__breakpointhook__
+        self.assertEqual(call_args, ((1, 2, 3), dict(four=4, five=5)))
+
+    def test_breakpoint_with_passthru_error(self):
+        def my_breakpointhook():
+            pass
+        try:
+            sys.breakpointhook = my_breakpointhook
+            self.assertRaises(TypeError, breakpoint, 1, 2, 3, four=4, five=5)
+        finally:
+            sys.breakpointhook = sys.__breakpointhook__
+
+    def test_envar_good_path_builtin(self):
```

Comment:
env var tests expect that sys.breakpointhook is the default hook. Maybe put these tests in a new test case and set sys.breakpointhook to sys.__breakpointhook__ in setUp() (and use addCleanup to restore the old value)?

Or maybe rewrite the tests to call directly sys.__breakpointhook__ rather than testing indirectly that breakpoint() calls sys.breakpointhook and expect sys.breakpointhook to be sys.__breakpointhook__?

Reply:
It's a good suggestion.  I think the clean-slate setup will address this.